File: Exavision.Seasense.Shared/Materials/Turret.cs
using Exavision.Seasense.Shared.Settings;

namespace Exavision.Seasense.Shared.Materials {
    public abstract class Turret<S, U> : Material<S, U>, ITurret where S : SettingMaterial, new() where U : IUnit, new() {
        public Turret(U unit) : base(unit) {

        }


        public override S GetSetting() {
            S setting = base.GetSetting();
            setting.MaterialType = MaterialType.Turret;
            return setting;
        }
    }
}

File: Exavision.Seasense/Client/src/app/materials/turret.ts
import { Capability } from "./capabilities/capability";
import { Factory } from "./factory";
import { Material } from "./material";
import { MaterialType } from "./material-type";
import { SettingCapability } from "./settings/setting-capability";
import { SettingMaterial } from "./settings/setting-material";

export class Turret implements Material {

  id: string;
  materials: Material[];
  capabilities: Capability[] = [];
  displayName: string;
  materialType: MaterialType;
  constructor(settingMaterial: SettingMaterial) {
    this.id = settingMaterial.id;
    this.materials = [];
    this.materialType = MaterialType.Turret;
    this.displayName = settingMaterial.displayName;
    let factory: Factory = new Factory();
    settingMaterial.materials.forEach((settingMaterialChild: SettingMaterial) => {
      let material: Material = factory.createMaterial(settingMaterialChild);
      this.materials.push(material);
    });
    settingMaterial.capabilities.forEach((settingCapability: SettingCapability) => {
      let capability: Capability = factory.createCapability(settingCapability);
      this.capabilities.push(capability);
    });
  }

}

